### PR TITLE
feat: add global withdraw limit logic

### PR DESCRIPTION
**Pull request by ramenforbreakfast**
We did not take the actual period tracking logic into account for the document. For the purposes of tracking logic, we need to store when the last time a period was reset, via `_lastGlobalPeriodReset` and `_lastUserPeriodReset`.

To make testing easier and cleaner, I have modified the limit so only one additional user needs to deposit the test amount to reach the global cap.

Originally I was going to submit both the global and user limit logic together, but the PR became too big and I have now split them up.

### CODE REVIEW

**Comment by ramenforbreakfast** [comment]
Closing due to decision to reduce code replication and have contracts inherit from a shared DepositRecordHook

File: apps/smart-contracts/core/contracts/WithdrawHook.sol [hunk 1]
```diff
@@ -45,6 +46,21 @@ contract WithdrawHook is IWithdrawHook, SafeAccessControlEnumerable {
...
+    ) {
+      _lastGlobalPeriodReset = block.timestamp;
+      _globalAmountWithdrawnThisPeriod = 0;
+    }
+    require(
+      _globalAmountWithdrawnThisPeriod + _amountAfterFee <=
+        _globalWithdrawLimitPerPeriod,
+      "global withdraw limit exceeded"
+    );
+    _globalAmountWithdrawnThisPeriod += _amountAfterFee;
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/WithdrawHook.sol:63]: it's inefficient to have a case that will only ever be used once
we can also reduce the number of operations in the reset case
and don't forget to change to use `_amountBeforeFee` instead
```suggestion
    if (
      _lastGlobalPeriodReset + _globalPeriodLength < block.timestamp
    ) {
      _lastGlobalPeriodReset = block.timestamp;
      _globalAmountWithdrawnThisPeriod = _amountBeforeFee;
    } else {
      require(
        _globalAmountWithdrawnThisPeriod + _amountBeforeFee <=
          _globalWithdrawLimitPerPeriod,
        "global withdraw limit exceeded"
      );
      _globalAmountWithdrawnThisPeriod += _amountBeforeFee;
    }
```

File: apps/smart-contracts/core/test/WithdrawHook.test.ts [hunk 2]
```diff
@@ -134,6 +144,157 @@ describe('=> WithdrawHook', () => {
...
+        await withdrawHook
+          .connect(vault)
+          .hook(user.address, TEST_AMOUNT_BEFORE_FEE, TEST_AMOUNT_AFTER_FEE)
+
+        expect(await withdrawHook.getLastGlobalPeriodReset()).to.eq(
+          await getLastTimestamp(ethers.provider)
+        )
+      })
+
+      it('sets last global reset to current time if global period has passed (>)', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/WithdrawHook.test.ts:161]: nit:
```suggestion
      it('sets last global reset to current time if global period passed', async () => {
```
similarly for others

File: apps/smart-contracts/core/test/WithdrawHook.test.ts [hunk 3]
```diff
@@ -134,6 +144,157 @@ describe('=> WithdrawHook', () => {
...
+        await withdrawHook
+          .connect(vault)
+          .hook(user.address, differentAmountToWithdraw, differentAmountToWithdraw)
+
+        expect(await withdrawHook.getGlobalAmountWithdrawnThisPeriod()).to.eq(
+          differentAmountToWithdraw
+        )
+      })
+
+      it("doesn't update last global reset if global period was just reached (==)", async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/WithdrawHook.test.ts:203]: nit:
```suggestion
      it("doesn't update last global reset if global period exactly reached", async () => {
```

similarly for others

File: apps/smart-contracts/core/test/WithdrawHook.test.ts [hunk 4]
```diff
@@ -134,6 +144,157 @@ describe('=> WithdrawHook', () => {
...
+          ethers.provider,
+          previousResetTimestamp + TEST_GLOBAL_PERIOD_LENGTH + 1
+        )
+
+        await withdrawHook
+          .connect(vault)
+          .hook(user.address, TEST_AMOUNT_BEFORE_FEE, TEST_AMOUNT_AFTER_FEE)
+
+        const currentResetTimestamp = await getLastTimestamp(ethers.provider)
+        expect(await withdrawHook.getLastGlobalPeriodReset()).to.eq(currentResetTimestamp)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/WithdrawHook.test.ts:177]: also assert that it's greater than the previous timestamp

File: apps/smart-contracts/core/test/WithdrawHook.test.ts [hunk 5]
```diff
@@ -134,6 +144,157 @@ describe('=> WithdrawHook', () => {
...
+        await withdrawHook
+          .connect(vault)
+          .hook(user.address, TEST_AMOUNT_BEFORE_FEE, TEST_AMOUNT_AFTER_FEE)
+
+        expect(await withdrawHook.getGlobalAmountWithdrawnThisPeriod()).to.eq(
+          previousGlobalAmountWithdrawn.add(TEST_AMOUNT_AFTER_FEE)
+        )
+      })
+
+      it('reverts if global withdraw limit exceeded during a period', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/WithdrawHook.test.ts:275]: nit: 
```suggestion
      it('reverts if global withdraw limit exceeded for period', async () => {
```

Comment (XavierEkkel) [inline apps/smart-contracts/core/test/WithdrawHook.test.ts:275]: what about a test for when it reaches the limit exactly?

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

**Pull request merged**